lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp

```diff
@@ -619,6 +877,130 @@ removeLocalNonNestedAccess(const AccessConflictAnalysis::Result &result,
   return changed;
 }
 
+// TODO: support multi-end access cases
+static EndAccessInst *getSingleEndAccess(BeginAccessInst *inst) {
+  EndAccessInst *end = nullptr;
+  for (auto *currEnd : inst->getEndAccesses()) {
+    if (end == nullptr)
+      end = currEnd;
+    else
+      return nullptr;
+  }
+  return end;
+}
+
+static void mergeEndAccesses(BeginAccessInst *parentIns,
+                             BeginAccessInst *childIns) {
+  auto *endP = getSingleEndAccess(parentIns);
+  if (!endP)
+    llvm_unreachable("not supported");
+  auto *endC = getSingleEndAccess(childIns);
+  if (!endC)
+    llvm_unreachable("not supported");
+
+  endC->setOperand(parentIns);
+  endP->eraseFromParent();
+}
+
+static bool canMergeEnd(BeginAccessInst *parentIns, BeginAccessInst *childIns) {
+  auto *endP = getSingleEndAccess(parentIns);
+  if (!endP)
+    return false;
+
+  auto *endC = getSingleEndAccess(childIns);
+  if (!endC)
+    return false;
+
+  return true;
+}
+
+// TODO: support other merge patterns
+static bool canMergeBegin(DominanceInfo *domTree,
+                          PostDominanceInfo *postDomTree,
+                          BeginAccessInst *parentIns,
+                          BeginAccessInst *childIns) {
+  if (!domTree->properlyDominates(parentIns, childIns)) {
```

Comment:
Shouldn't this be an assert. How can an non-dominated access end up in the merge map?

Reply:
it should. we don't even need domTree just for the assert, wrote that code too late at night :)